### PR TITLE
Enhanced MMapIndexedDataset: less memory, higher speed

### PR DESCRIPTION
**Pull request by davidecaroselli**
I have made an upgrade to my previous implementation of MMapIndexedDataset, now:
- It uses up to **4 times less memory and disk space**
- Words per second is slightly improved thanks to less memory access

### CODE REVIEW

**Comment by myleott** [comment]
Nice!

File: fairseq/data/indexed_dataset.py [hunk 1]
```diff
@@ -15,9 +15,16 @@
 from . import FairseqDataset
 
 
-def make_builder(out_file, impl):
+def __best_fitting_dtype(vocab_size=None):
+    if vocab_size is not None and vocab_size < 65500:
+        return np.uint16
+    else:
+        return np.int32
```
Comment (myleott) [inline fairseq/data/indexed_dataset.py]: This makes sense. Is this the main cause of the memory/speed improvement?

File: fairseq/data/indexed_dataset.py [hunk 2]
```diff
@@ -440,11 +448,11 @@ def __len__(self):
...
     def __getitem__(self, i):
         ptr, size = self._index[i]
-        tensor = torch.from_numpy(np.frombuffer(self._bin_buffer, dtype=self._index.dtype, count=size, offset=ptr))
-        if tensor.dtype == torch.int64:
-            return tensor
-        else:
-            return tensor.long()
+        np_array = np.frombuffer(self._bin_buffer, dtype=self._index.dtype, count=size, offset=ptr)
+        if self._index.dtype != np.int64:
+            np_array = np_array.astype(np.int64)
```
Comment (myleott) [inline fairseq/data/indexed_dataset.py]: Does this change improve memory usage and/or speed as well?

Comment (davidecaroselli) [reply fairseq/data/indexed_dataset.py]: This is due the fact that invoking `tensor.long()` on a `uint16` tensor was throwing an exception (tensor supports casting only on `int64`, `int32`, `uint8`, ... except `uint16`).
This way I bypass the problem!

File: fairseq/data/indexed_dataset.py (hunk 1 above)
Comment (davidecaroselli) [reply fairseq/data/indexed_dataset.py]: Yes, this is actual boost. I now have a model that is 1/4th of the original and this means a huge saving in memory, disk and also load time!

**Comment by davidecaroselli** [comment]
I'm running the very last test where I train a net on 8GPU for 6 hours and test that the translation results are equivalent to the same net trained with old version of MMapIndexedDataset.

PS: I just noticed the new version is out 0.7.0. That's great, but I was really hoping to have this improvement in the release! Too late I guess... Should I have to wait ~3 months, or are you planning to have a more regular and fast release schedule? It would be great to see all new improvements available in less time, even more important in this fast-growing stage of the framework

**Comment by myleott** [comment]
> Should I have to wait ~3 months, or are you planning to have a more regular and fast release schedule?

We'll try to release more regularly. I cut 0.7.0 because we have a major logging refactoring landing soon and I wanted to cut a release right before it. We'll probably move to 0.8.0 very soon once that's merged.

I can put this in a 0.7.1 as soon as it's merged.

**Comment by davidecaroselli** [comment]
Thanks @myleott for the amazing product! :)

**Comment by myleott** [comment]
0.7.1 is on pypi with this change!

**Comment by davidecaroselli** [comment]
Thanks a million!

**Pull request closed without merging**